errors/errors.go

```diff
@@ -67,79 +71,85 @@ func Wrap(err error, context ...map[string]string) *Error {
+func BadRequest(code, message string, params map[string]string) *Error {
+	return errorFactory(errCode(ErrBadRequest, code), message, params)
 }
 
 // BadResponse creates a new error representing a failure to response with a valid response
 // Examples of this would be a handler returning an invalid message format
-func BadResponse(message string, context ...map[string]string) *Error {
-	return errorFactory(ErrBadResponse, message, context...)
+func BadResponse(code, message string, params map[string]string) *Error {
+	return errorFactory(errCode(ErrBadResponse, code), message, params)
 }
 
 // Timeout creates a new error representing a timeout from client to server
-func Timeout(message string, context ...map[string]string) *Error {
-	return errorFactory(ErrTimeout, message, context...)
+func Timeout(code, message string, params map[string]string) *Error {
+	return errorFactory(errCode(ErrTimeout, code), message, params)
 }
 
 // NotFound creates a new error representing a resource that cannot be found. In some
 // cases this is not an error, and would be better represented by a zero length slice of elements
-func NotFound(message string, context ...map[string]string) *Error {
-	return errorFactory(ErrNotFound, message, context...)
+func NotFound(code, message string, params map[string]string) *Error {
+	return errorFactory(errCode(ErrNotFound, code), message, params)
 }
 
 // Forbidden creates a new error representing a resource that cannot be accessed with
 // the current authorisation credentials. The user may need authorising, or if authorised,
 // may not be permitted to perform this action
-func Forbidden(message string, context ...map[string]string) *Error {
-	return errorFactory(ErrForbidden, message, context...)
+func Forbidden(code, message string, params map[string]string) *Error {
+	return errorFactory(errCode(ErrForbidden, code), message, params)
 }
 
 // Unauthorized creates a new error indicating that authentication is required,
 // but has either failed or not been provided.
-func Unauthorized(message string, context ...map[string]string) *Error {
-	return errorFactory(ErrUnauthorized, message, context...)
+func Unauthorized(code, message string, params map[string]string) *Error {
+	return errorFactory(errCode(ErrUnauthorized, code), message, params)
 }
 
-// errorConstructor returns a `*Error` with the specified code, message and context. The main work
-// consists of managing the map[string]string's at the end of the arguments list.
-// In practice we only ever pass in two: the first one is public and will be sent to the client,
-// the second one is private and can contain internal information that is useful for debugging
-func errorFactory(code int, message string, context ...map[string]string) *Error {
+// errorConstructor returns a `*Error` with the specified code, message and params.
+// Builds a stack based on the current call stack
+func errorFactory(code string, message string, params map[string]string) *Error {
```

Comment:
Not sure if this is really bad design, but I would have made `params` optional so that we don't have to put `nil` at the end of the constructor every time we use it. FWIW `Sprintf` uses a similar trick